### ISSUE TITLE
Move to requiring setuptools

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7177]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Alex reported|
|-|-|
|Trac ID|trac#7177|
|Type|enhancement|
|Created|2014-04-16 13:49:18Z|
|Branch|https://github.com/twisted/twisted/tree/setuptools-7177|

`setuptools` is, practically speaking, available on every installation platform you care about, since `pip` makes it happen.

<details><summary>Searchable metadata</summary>

```
trac-id__7177 7177
type__enhancement enhancement
reporter__Alex Alex
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_setuptools_7177 branches/setuptools-7177
branch_author__ralphm__hawkowl ralphm, hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1397656158000000 1397656158000000
changetime__1439206436905917 1439206436905917
version__None None
owner__hawkowl hawkowl
cc__adiroiban
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|Alex commented|
|-|-|

This is resolved by https://github.com/twisted/twisted/pull/21

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_2"></a>|@ralphm set owner to Alex|
|-|-|

This mostly looks good. However, the news file doesn't conform to our standard. It is meant to be a high level description of what happened, with the subject being the module that changed and a verb prefixed by _now_. In this case it would need to be something like:

```
Twisted now depends on ...
```

You probably don't want to mention 'setup.py' at all.

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|Alex set owner to @ralphm|
|-|-|

The PR has been updated to address this. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_4"></a>|@ralphm commented|
|-|-|

(In [42434]) Branching to setuptools-six-7177.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_5"></a>|@ralphm set owner to Alex|
|-|-|

The branch is fine now, but it turns out that it must block on [#3696](https://github.com/twisted/twisted/issues/3696) (which [#7092](https://github.com/twisted/twisted/issues/7092) was marked a duplicate of), because after/while landing, six would need to be installed on all our build slaves.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_6"></a>|@radix commented|
|-|-|

I'm not sure why this must block on [#3696](https://github.com/twisted/twisted/issues/3696) -- if we depend on six, we depend on it strictly, not optionally.

I think the only thing blocking it is somehow running on the buildslaves -- whether by updating our build steps to automatically install dependencies, or by manually installing six on all of them.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_7"></a>|@ralphm commented|
|-|-|

I think [#7092](https://github.com/twisted/twisted/issues/7092) might have been a better description, but glyph made that a dupe of [#3696](https://github.com/twisted/twisted/issues/3696) when we discussed this problem. Feel free to undupe. Manually installing stuff is a bad idea. The patch itself is clearly making it a hard dependency.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

(In [45377]) Branching to setuptools-7177.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl removed owner|
|-|-|

I don't think we need six until we have something actually using it. Otherwise, this is getting in a shape. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

(builders spun, also)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Please update the title of this ticket. It says 'Add six as a dependency ... ' but no six is added here. The description is fine as it does not talk about six.

as long as we continue to duplicate six code into twisted.python.compat we don't need six... and since we don't depend on six people will continue to duplicate code in compat rather than using six.

I guess that this ticket/branch is only about the migration to setuptools.

-----

I think we should document that Twisted depends on setuptools, at leaset in the INSTALL file. 

As far as I know, pip can also work with Distribute.

I see that in 'very old' versions of setuptools there used to be a bug with install_requires.

I am ok with not handling that error in the code but I think that we should document the version required for Twisted.


Please address my comments and merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl commented|
|-|-|

Hi Adi, thanks for the review.

- Ticket title changed.
- I couldn't find a version of setuptools that exhibited that bug  -- I asked lifeless and he hadn't heard of such a bug existing, so I tried the oldest ones I could on PyPI and they all worked fine.
- I'll add the requirement for setuptools.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_15"></a>|@hawkowl set status to `closed`|
|-|-|

(In [45435]) Merge setuptools-7177: Require setuptools for installation.

Author: Alex, hawkowl
Reviewers: ralphm, adiroiban
Fixes: [#7177](https://github.com/twisted/twisted/issues/7177)
